diff -Naur linux-4.19.26.orig/arch/mips/include/asm/mach-generic/kmalloc.h linux-4.19.26/arch/mips/include/asm/mach-generic/kmalloc.h
--- linux-4.19.26.orig/arch/mips/include/asm/mach-generic/kmalloc.h	2019-02-27 12:09:03.000000000 +0300
+++ linux-4.19.26/arch/mips/include/asm/mach-generic/kmalloc.h	2019-03-04 12:40:11.022829784 +0300
@@ -7,7 +7,7 @@
  * Total overkill for most systems but need as a safe default.
  * Set this one if any device in the system might do non-coherent DMA.
  */
-#define ARCH_DMA_MINALIGN	128
+#define ARCH_DMA_MINALIGN	16 /* 16 byte cache line on stb225 */
 #endif
 
 #endif /* __ASM_MACH_GENERIC_KMALLOC_H */
diff -Naur linux-4.19.26.orig/arch/mips/include/asm/mach-pnx833x/pnx833x.h linux-4.19.26/arch/mips/include/asm/mach-pnx833x/pnx833x.h
--- linux-4.19.26.orig/arch/mips/include/asm/mach-pnx833x/pnx833x.h	2019-02-27 12:09:03.000000000 +0300
+++ linux-4.19.26/arch/mips/include/asm/mach-pnx833x/pnx833x.h	2019-03-04 12:40:11.022829784 +0300
@@ -199,4 +199,8 @@
 #define PNX8335_SATA_MODULE_ID_MODULE_ID_SHIFT	16
 #define PNX8335_SATA_MODULE_ID_VALUE		0xA099
 
+#define PNX8335_SIF_CPU_OPTIMIZATION			PNX833X_REG(0x3038)
+#define PNX8335_SIF_CPU_OPTIMIZATION_CPU_FREE_MASK	0x1
+#define PNX8335_SIF_CPU_OPTIMIZATION_CPU_FREE_SHIFT	0
+
 #endif
diff -Naur linux-4.19.26.orig/arch/mips/Kconfig linux-4.19.26/arch/mips/Kconfig
--- linux-4.19.26.orig/arch/mips/Kconfig	2019-02-27 12:09:03.000000000 +0300
+++ linux-4.19.26/arch/mips/Kconfig	2019-03-04 12:40:11.026829744 +0300
@@ -1276,6 +1276,8 @@
 config SOC_PNX8335
 	bool
 	select SOC_PNX833X
+	select MIPS_L1_CACHE_SHIFT_4
+	select USB_ARCH_HAS_HCD
 
 config MIPS_SPRAM
 	bool
diff -Naur linux-4.19.26.orig/arch/mips/pnx833x/common/setup.c linux-4.19.26/arch/mips/pnx833x/common/setup.c
--- linux-4.19.26.orig/arch/mips/pnx833x/common/setup.c	2019-02-27 12:09:03.000000000 +0300
+++ linux-4.19.26/arch/mips/pnx833x/common/setup.c	2019-03-04 12:40:11.026829744 +0300
@@ -38,9 +38,9 @@
 
 int __init plat_mem_setup(void)
 {
-	/* set mips clock to 320MHz */
+	/* set mips clock to 400MHz */
 #if defined(CONFIG_SOC_PNX8335)
-	PNX8335_WRITEFIELD(0x17, CLOCK_PLL_CPU_CTL, FREQ);
+	PNX8335_WRITEFIELD(0x1F, CLOCK_PLL_CPU_CTL, FREQ);
 #endif
 	pnx833x_gpio_init();	/* so it will be ready in board_setup() */
 
@@ -57,5 +57,10 @@
 	iomem_resource.start = 0;
 	iomem_resource.end = ~0;
 
+#if defined(CONFIG_SOC_PNX8335)
+	/* Allows full bandwidth for CPU when no AGU request is pending */
+	PNX8335_WRITEFIELD(0x1, SIF_CPU_OPTIMIZATION, CPU_FREE);
+#endif
+
 	return 0;
 }
